src/coreclr/jit/assertionprop.cpp

```diff
@@ -4257,6 +4257,18 @@ GenTree* Compiler::optAssertionPropLocal_RelOp(ASSERT_VALARG_TP assertions, GenT
     return optAssertionProp_Update(op2, tree, stmt);
 }
 
+/*****************************************************************************
+ *
+ *  Given a tree ConditionalOp, check using standard morph function.
+ */
+
+GenTree* Compiler::optAssertionProp_ConditionalOp(ASSERT_VALARG_TP assertions, GenTree* tree, Statement* stmt)
+{
+    assert(tree->OperIsConditional());
+    GenTree* newTree = fgMorphTree(tree);
+    return optAssertionProp_Update(newTree, tree, stmt);
```

Comment:
It's not right to just reinvoke morph in assertion propagation. If you expect constant folding to take place, it should be done as regular part of morph (and I see the code for that already exists).

Additionally, it does not make sense w.r.t. the current phase order, since if conversion runs after assertion propagation.

Reply:
>Additionally, it does not make sense w.r.t. the current phase order,

That's due to If Conversion originally running before assertion propagation. I wasn't sure if it was still best to have some code to handle SELECTs here - but without it it'll just assert in the default case, which is probably the correct behaviour.
I'll remove all my changes from assertionprop.cpp.